src/sagas/vmChanges.js

```diff
@@ -175,56 +165,194 @@ function* composeProvisionSourceTemplate ({ vm, basic, disks }) {
+        (changed, index) => ({
+          disk: changed || { id: disksFromTemplate[index].disk.id },
+        })
+      ),
+    }
+  }
 
-      // make sure disk format (and therefore sparse) matches the template's storageAllocation
-      if (templateStorageAllocation === 'thin') {
-        changesToTemplateDisk.format = 'cow'
-      } else if (templateStorageAllocation === 'clone') {
-        changesToTemplateDisk.format = vmStorageAllocation === 'thin' ? 'cow' : 'raw'
-      }
+  return [vmUpdates, storageAllocation === 'clone']
+}
 
-      // did the storage domain change?
-      if (disk.storageDomainId !== templateDisk.get('storageDomainId')) {
-        diskRequiresClone = true
-        changesToTemplateDisk.storage_domains = {
-          storage_domain: [{ id: disk.storageDomainId }],
-        }
-      }
+/**
+ * Select the disk's __format__ and __sparse__ based on where the disk is coming from,
+ * what changes the user has made to the VM from the template, and the disk's target
+ * storage domain.
+ *
+ * Template values and Webadmin choices in the new VM modal flow down ultimately to the
+ * template defined VM disk's specific attributes.  The default 'storage allocation' maps
+ * based on the template optimizedFor, VM optimizedFor and target storage domain.  This
+ * chart captures what format and sparse values are produced in webadmin:
+ *
+ *   [Template OptimizedFor -> VM OptimizedFor -> SD changes? ~~ Storage allocation ... disk format/sparse outcome]
+ *   (* = default)
+ *
+ *   Server  -> Server*  -> No SD changes* ~~ Clone ... format stays the same, sparse stays the same
+ *   Server  -> Server*  -> SD changes     ~~ Clone ... format stays the same, sparse changes as necessary based on format + SD
+ *   Server  -> Desktop  -> No SD changes* ~~ Thin  ... format: cow, sparse: true
+ *   Server  -> Desktop  -> SD changes     ~~ Clone ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Desktop* -> No SD changes* ~~ Thin  ... format: cow, sparse: true
+ *   Desktop -> Desktop* -> SD changes     ~~ Clone ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Server   -> No SD changes* ~~ Clone ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Server   -> SD changes     ~~ Clone ... format: cow, sparse: true (Thin settings retained)
+ *
+ * @param userDisk Disk as approved by the user in the Create VM Storage step
+ * @param baseDisk Disk as defined in the template
+ * @param template Template the VM is being created from
+ * @param templateIsDesktop Is the template type 'desktop'?
+ * @param vmIsDesktop Is the VM being created with user selectable type/optimizedFor 'desktop'?
+ * @param vmStorageAllocationIsThin Is the net storage allocation for this VM 'thin'?
+ * @param targetSD The storage domain where the disk will be created
+ * @param targetCluster The cluster where the VM will be created
+ */
+function determineTemplateDiskFormatAndSparse (
```

Comment:
>> can we unit test according to scenarios above?
>
> Probably.

We really should.

Reply:
I haven't addressed this unit test yet.  It's just painful to think of the permutations.

Reply:
I added some unit tests specifically for `determineTemplateDiskFormatAndSparse()` that follow the data flow not necessarily all input permutations..